src/common/ceph_json.cc

```diff
@@ -249,7 +249,12 @@ bool JSONParser::parse(const char *buf_, int len)
       if (data.type() == str_type) {
         val.set(data.get_str(), true);
       } else {
-        val.set(json_spirit::write_string(data), false);
+        std::string_view s = json_spirit::write_string(data);
+        if (s.size() == (uint64_t)len) { /* Check if entire string is read */
+          val.set(s, false);
+        } else {
+          set_failure();
+        }
```

Comment:
in the example, `"312CNV15CC2ZN44IPB24A"` was being parsed as integer `312` instead of the full string. doesn't this change just cause the parsing to fail, instead of correctly parsing it as a string?

Reply:
I am not sure if its only string type (though likely) which gets misinterpreted as integer. I haven't tested other object types. So setting the value as string here dint seem right. Also data.type() gets set in json_spirit library and cannot be changed to str_type here IIUC. It may mislead if the caller checks for data.type().

By returning failure here, in "JSONFormattable::set()", 

  if (is_valid_json) {
    f->decode_json(&jp);
  } else {
    f->type = FMT_VALUE; 
    f->value.set(val, !is_numeric(val));                                                       

^^^^ The value here is again checked if actual numerical or not before setting the value as string.

Reply:
@cbodley .. request to review .. fixing this issue might help to configure cloud tier storage class (with default access keys) in teuthology tests.

Reply:
ok, from your comment above it sounds like this will do the right thing. can you please add a simple test case to `test/common/test_json_formatter.cc` to cover this?

it's still hard to be confident that nothing in ceph is relying on this behavior, but i don't mind merging this now that quincy is forked. we'll have a whole release cycle to uncover any issues